sourcing/domain/service.go

```diff
@@ -210,78 +270,76 @@ func getItemIdsWithProduct(dc *decorator.DecoratedCart, product domain.BasicProd
 			result = append(result, ItemID(di.Item.ID))
 		}
 	}
+
 	return result
 }
 
-// allocateItem returns the itemAllocation and the remaining stock for the given item.
-// The passed productSourcestock is used - and the remaining productSourcestock is returned. In case a source is not yet given in productSourcestock it will be fetched
-func (d *DefaultSourcingService) allocateItem(ctx context.Context, productSourcestock map[string]map[Source]int, decoratedItem decorator.DecoratedCartItem, deliveryInfo cartDomain.DeliveryInfo) (ItemAllocation, map[string]map[Source]int) {
-	var resultItemAllocation = ItemAllocation{
-		AllocatedQtys: make(AllocatedQtys),
-	}
-	// copy given known stock
-	remainingSourcestock := productSourcestock
-
-	productID := decoratedItem.Product.GetIdentifier()
+func (d *DefaultSourcingService) allocateItem(ctx context.Context, productSourcestock map[string]map[Source]int, product domain.BasicProduct, qtyToAllocate int, deliveryInfo cartDomain.DeliveryInfo) (AllocatedQtys, map[string]map[Source]int, error) {
+	productID := product.GetIdentifier()
```

Comment:
I guess you don't need productID here anymore. You can validate product.GetIdentifier() directly and then not pass product id to d.allocateFromSources()